---
title: "Deconvolution with `SPOTlight`"
date: "`r BiocStyle::doc_date()`"
author:
- name: Marc Elosua-Bayes
  email: marc.elosua@cnag.crg.eu
- name: Helena L. Crowell
  affiliation:
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - &SIB SIB Swiss Institute of Bioinformatics, University of Zurich, Switzerland
  email: helena.crowell@uzh.ch
abstract: >
    Text...
package: "`r BiocStyle::pkg_ver('SPOTlight')`"
bibliography: "`r file.path(system.file('extdata', package = 'SPOTlight'), 'refs.bib')`"
vignette: >
  %\VignetteIndexEntry{"SPOTlight"}
  %\VignettePackage{SPOTlight}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

The goal of `SPOTlight`is to provide a tool that enables the deconvolution of 
cell types and cell type proportions present within each capture location 
comprising mixtures of cells. Originally developed for 10X's Visium - spatial
transcriptomics - technology, it can be used for all technologies returning
mixtures of cells. `SPOTlight` is based on learning topic profile signatures,
by means of an NMFreg model, for each cell type and finding which combination of
cell types fits best the spot we want to deconvolute. Find below a graphical
abstract visually summarizing the key steps.


![](../inst/extdata/schematic.png)

# Workflow

```{r}
library(SPOTlight)
x <- .mock_sc()
y <- .mock_sp(x)
```

## Preprocessing

If the dataset is very large we want to downsample it to train the model, 
both in of number of cells and number of genes. To do this, we want to keep
a representative amount of cells per cluster and the most important genes. 

In the paper we show how downsampling the number of cells per cell type to ~100
doesn't affect the performance of the model since performance metrics show
marginal improvement for larger cell numbers while computational time and
resources increase greatly. Furthermore, restricting the gene set to the marker
genes for each cell type along with up to 3.000 highly variable genes further
optimizes performance and computational resources. You can find a more detailed
explanation in the original paper [here]().

# TODO - add link to paper

### Feature selection
Our first step is to get the marker genes for each cell type. You can use
whichever method you want as long as it returns a weight indicating the importance
of that gene for that cell type. Examples can be `avgLogFC`, `AUC`,
`pct.expressed`..., 
```{r}
z <- .get_mgs(x = x, n_top = 100)
z
```

### Cell Downsampling

Next you randomly select up to 100 cells/cell-type. If a cell type
is comprised of <100 cells than all the cells will be used. If we have very
biologically different cell types (B cells vs T cells vs Macrophages vs
Epithelial) we can use fewer cells since their transcriptional profiles will 
be very different. In cases when we have more transcriptionally similar cell
types we then need to increase our N to capture the biological heterogeneity
between them.

```{r}
n_cell <- 20
x_ls <- split(colData(x), x$type)
id_keep <- lapply(x_ls, function(.) {
  sample(
    x = rownames(.),
    size = ifelse(n_cell > nrow(.), nrow(.), n_cell),
    replace = FALSE)
  })

x <- x[, unlist(id_keep)]
```

## Deconvolution

You are now set to run `SPOTlight` to deconvolute the spots! Briefly 

# Get SPOTlight explanation from paper

You can visualize the above explanation in the following workflow scheme:

![](../inst/extdata/workflow.png)

```{r}
res <- SPOTlight(x, y, 
    groups = x$type,
    mgs = z, 
    group_id = "type")
```

```{r}
head(mat <- res[[2]])
err <- mat[,  ncol(mat)]
mat <- mat[, -ncol(mat)]
```

# Visualization

## Topic profiles

```{r}

```

## Co-localization

Now that we know which cell types are found within each spot we can make a graph 
representing spatial interactions where cell types will have stronger edges 
between them the more often we find them within the same spot. 

See [here](https://www.r-graph-gallery.com/network.html) for additional graphical parameters.

```{r}
plotInteractions(mat, "heatmap")
plotInteractions(mat, "network")
```

## Scatterpie

```{r}

```

# Session information

```{r session-info}
sessionInfo()
```

# References
